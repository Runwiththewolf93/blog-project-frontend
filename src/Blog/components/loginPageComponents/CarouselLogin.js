import { Carousel, Image } from "react-bootstrap";
import useUnsplashImages from "../../hooks/useUnsplash";

/**
 * Renders a carousel login component.
 *
 * @param {number} loginFormHeight - The height of the login form.
 * @return {JSX.Element} The rendered carousel login component.
 */
// CarouselLogin component
const CarouselLogin = ({ loginFormHeight }) => {
  const images = useUnsplashImages("blog", 3) || [];

  const carouselText = [
    {
      h3: "Stay up to date",
      p: "With the many topics available for browsing on our website",
    },
    {
      h3: "Spark your interest",
      p: "Get your creative juices flowing with our many social networks",
    },
    {
      h3: "Get relevant information",
      p: "For your vacation trip to another country around the world",
    },
  ];

  return (
    <Carousel fade pause="hover" controls={false} data-testid="carousel">
      {images.map((image, index) => {
        // Check if image has the required properties
        if (!image.id || !image.urls?.regular || !image.alt_description) {
          return null;
        }

        return (
          <Carousel.Item
            interval={10000}
            key={image.id}
            data-testid="carousel-item"
          >
            <Image
              style={{ objectFit: "cover", height: loginFormHeight || "500px" }}
              src={image.urls?.regular}
              alt={image.alt_description}
              rounded
            />
            <Carousel.Caption data-testid="carousel-caption">
              <h3 className="text-black">
                <strong>{carouselText[index].h3}</strong>
              </h3>
              <p className="text-black">
                <strong>{carouselText[index].p}</strong>
              </p>
            </Carousel.Caption>
          </Carousel.Item>
        );
      })}
    </Carousel>
  );
};

export default CarouselLogin;
