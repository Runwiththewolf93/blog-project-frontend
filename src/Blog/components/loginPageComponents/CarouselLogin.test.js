import { render, screen } from "@testing-library/react";
import CarouselLogin from "./CarouselLogin";
import useUnsplashImages from "../../hooks/useUnsplash";

jest.mock("../../hooks/useUnsplash", () => jest.fn());

// Mock data for the hook
const mockImages = [
  {
    id: "1",
    urls: { regular: "https://example.com/image1.jpg" },
    alt_description: "Image 1",
  },
  {
    id: "2",
    urls: { regular: "https://example.com/image2.jpg" },
    alt_description: "Image 2",
  },
  {
    id: "3",
    urls: { regular: "https://example.com/image3.jpg" },
    alt_description: "Image 3",
  },
];

// Mock data for the loginFormHeight prop
const mockLoginFormHeight = "500px";

describe("CarouselLogin_function", () => {
  // Tests that the function renders a carousel login component with images and captions
  it("test_happy_path_renders_carousel_login_component", () => {
    useUnsplashImages.mockReturnValue(mockImages);

    render(<CarouselLogin loginFormHeight={mockLoginFormHeight} />);

    const carousel = screen.getByTestId("carousel");

    expect(carousel).toBeInTheDocument();
  });

  // Tests that images and captions are displayed in the carousel
  it("test_happy_path_images_and_captions_displayed_in_carousel", () => {
    useUnsplashImages.mockReturnValue(mockImages);

    render(<CarouselLogin loginFormHeight={mockLoginFormHeight} />);

    const carouselItems = screen.getAllByTestId("carousel-item");
    expect(carouselItems.length).toBe(3);

    const carouselCaptions = screen.getAllByTestId("carousel-caption");
    expect(carouselCaptions.length).toBe(3);
  });

  // Tests that the carousel has fade effect and pause on hover
  it("test_happy_path_carousel_has_fade_effect_and_pause_on_hover", () => {
    useUnsplashImages.mockReturnValue(mockImages);

    render(<CarouselLogin loginFormHeight={mockLoginFormHeight} />);

    const { asFragment } = render(
      <CarouselLogin loginFormHeight={mockLoginFormHeight} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  // Tests that the function handles the edge case where no images are returned from useUnsplashImages hook
  it("test_edge_case_no_images_returned_from_useUnsplashImages_hook", () => {
    useUnsplashImages.mockReturnValue([]);

    render(<CarouselLogin loginFormHeight={mockLoginFormHeight} />);

    const carouselItems = screen.queryAllByTestId("carousel-item");
    expect(carouselItems.length).toBe(0);
  });

  // Tests that the function handles the edge case where images returned from useUnsplashImages hook are not in expected format
  it("test_edge_case_images_returned_from_useUnsplashImages_hook_not_in_expected_format", () => {
    useUnsplashImages.mockReturnValue([{ id: "1" }]);

    render(<CarouselLogin loginFormHeight={mockLoginFormHeight} />);

    const carouselItems = screen.queryAllByTestId("carousel-item");
    expect(carouselItems.length).toBe(0);
  });

  // Tests that the function handles the edge case where loginFormHeight prop is not provided or not a number
  it("test_edge_case_loginFormHeight_prop_not_provided_or_not_a_number", () => {
    useUnsplashImages.mockReturnValue(mockImages);

    // Render without loginFormHeight
    render(<CarouselLogin />);
    let carouselImages = screen.getAllByRole("img");
    carouselImages.forEach(img => {
      // Check if the height style is the default value
      expect(img.style.height).toBe("500px");
    });

    // Render with loginFormHeight
    render(<CarouselLogin loginFormHeight={mockLoginFormHeight} />);
    carouselImages = screen.getAllByRole("img");
    carouselImages.forEach(img => {
      // Check if the height style is equal to loginFormHeight
      expect(img.style.height).toBe(mockLoginFormHeight);
    });
  });
});
