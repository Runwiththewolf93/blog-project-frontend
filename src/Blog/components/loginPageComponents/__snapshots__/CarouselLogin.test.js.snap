// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselLogin_function test_happy_path_carousel_has_fade_effect_and_pause_on_hover 1`] = `
<DocumentFragment>
  <div
    class="carousel slide carousel-fade"
    data-testid="carousel"
  >
    <div
      class="carousel-indicators"
    >
      <button
        aria-current="true"
        aria-label="Slide 1"
        class="active"
        data-bs-target=""
        type="button"
      />
      <button
        aria-current="false"
        aria-label="Slide 2"
        data-bs-target=""
        type="button"
      />
      <button
        aria-current="false"
        aria-label="Slide 3"
        data-bs-target=""
        type="button"
      />
    </div>
    <div
      class="carousel-inner"
    >
      <div
        class="active carousel-item"
        data-testid="carousel-item"
        interval="10000"
      >
        <img
          alt="Image 1"
          class="rounded"
          src="https://example.com/image1.jpg"
          style="object-fit: cover; height: 500px;"
        />
        <div
          class="carousel-caption"
          data-testid="carousel-caption"
        >
          <h3
            class="text-black"
          >
            <strong>
              Stay up to date
            </strong>
          </h3>
          <p
            class="text-black"
          >
            <strong>
              With the many topics available for browsing on our website
            </strong>
          </p>
        </div>
      </div>
      <div
        class="carousel-item"
        data-testid="carousel-item"
        interval="10000"
      >
        <img
          alt="Image 2"
          class="rounded"
          src="https://example.com/image2.jpg"
          style="object-fit: cover; height: 500px;"
        />
        <div
          class="carousel-caption"
          data-testid="carousel-caption"
        >
          <h3
            class="text-black"
          >
            <strong>
              Spark your interest
            </strong>
          </h3>
          <p
            class="text-black"
          >
            <strong>
              Get your creative juices flowing with our many social networks
            </strong>
          </p>
        </div>
      </div>
      <div
        class="carousel-item"
        data-testid="carousel-item"
        interval="10000"
      >
        <img
          alt="Image 3"
          class="rounded"
          src="https://example.com/image3.jpg"
          style="object-fit: cover; height: 500px;"
        />
        <div
          class="carousel-caption"
          data-testid="carousel-caption"
        >
          <h3
            class="text-black"
          >
            <strong>
              Get relevant information
            </strong>
          </h3>
          <p
            class="text-black"
          >
            <strong>
              For your vacation trip to another country around the world
            </strong>
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
